Reference implementation of BIP135 (generalized versionbits)

sanch0panza opened the pull request:
This is an initial reference implementation of my 'generalized versionbits' BIP proposal, the draft specification of which you can find at this link:
https://github.com/bitcoin/bips/blob/master/bip-0135.mediawiki

It provides flexibly configurable parameters for each of the 29 BIP9 deployment bits, including the possibility of

- different thresholds / window sizes
- customizable grace periods (based either on minimum number of blocks or minimum time in seconds elapsed since lock-in)
- explicitly agnostic to fork type (not only intended for soft forks)

The configuration of the deployment bits can be read in from a comma-separated values file, and using the `-dumpforks` option, the built-in configuration can also be dumped out in CSV format to standard output.

BIP135 aims to be completely backward compatible (if suitably configured) with the BIP9 state machine. I have taken care to ensure this by leaving the original tests as unchanged as possible, and adding tests for the new functionality in separate scripts / programs.

The "unknown versions" alerting logic in main.cpp needed to be replaced, which slightly alters its behaviour. It now alerts based on a flat > 50/100 threshold and window, and does not attempt to divine the activation states of unknown signals. It also reports on first detection of an unknown signal, as well as gradated (25%, 50%, 70%, 90%, 95%) crossing thresholds, and loss of previously detected unknown signals during a 100-block window.
The 'unknown version' check itself - performed against the computed version according to Unlimited code - should be backward compatible. The notification / alerting code has only been lightly tested, as it is basically optional (not seen as part of core BIP135 requirements) and probably should be implemented in a fancier way, to allow more user control over when/what/how it reports.

In some places this reference implementation is surely a little worse technically than it should be, since I'm not an expert C++ programmer. Please feel free to criticize and offer suggestions for improvement. I'll do my best to provide rationale about any aspects of the implementation.

---

Update: This is a fresh PR created based on https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/458 which experienced problems updating in github while attempting to rebase on BU 1.3.0.x.

sanch0panza committed:
Add a new Python regression test, bipgenvbvoting-forks (adapted from bip9-softforks)

This uses a new getblockchaininfo RPC call section, 'bipgenvb_forks'.
This is a temporary name, which can be adapted if this BIP receives a number of its own (and the bip9_softforks section can be deprecated at such time, as the new section provides a superset of the information).

sanch0panza committed:
Add function to return the path to the deployment configuration file (FORKS_CSV_FILENAME)

This file does not need to be present at runtime - if not, the built-in defaults are used.
If the file is found (either at the default location or that specified by `-forks=<filepath>` command line option), then the file is parsed.

By default, the client is deployed using its built-in defaults.

sanch0panza committed:
Example forks.csv file (for reference only)

This file is provided only as an example of a complete deployment configuration file.
It is not deployed or used in any way at the moment.

If a user wishes to obtain an accurate representation of the client's built-in
deployment data, they may run using the `-dumpforks` command line option,
which will produce on standard output something similar to this file, which can
be redirected to a forks.csv file to be adapted and used.

sanch0panza committed:
Deployment configuration (CSV) file reading module and unit tests for it

NOTE: This depends on additional data structures added in subsequent commit.

sanch0panza committed:
Add new deployment members to consensus parameters, adapt the chainParams for all networks

Removed the old global parameters (nRuleChangeActivationThreshold, nMinerConfirmationWindow).
The new per-bit parameters are preconfigured to remain compatible with existing deployments.

Additional functions:
- ModifiableParams() returns modifiable chain params (for update by CSV file parsing routine).
- isConfiguredDeployment() is a helper function which reports if a fork bit considered "configured". This is used by the main logic to decide which fork bits to process further using the conditions checker.
- NetworkDeploymentInfoCSV() provides a network's deployment data as a string which can be streamed to output (used by -dumpforks option).

sanch0panza committed:
Update existing versionbits_tests.cpp

The only changes here are needed to
a) make it work given that we will update the AbstractThresholdConditionChecker (additional member functions provided in a subsequent commit)
b) a fix to the disjointness check, which previously worked on the unwritten assumption that the deployments would be ordered in a way that would reduce effort needed to check disjointness. With the opening up of the configuration interface, and even before then, this is a bad assumption, so the check was generalized to compare each fork against all the others.

No changes to the rest of the test's logic have been made, to ensure it still tests correctly, to fulfil the backward compatibility goal of bip-genvbvoting.

sanch0panza committed:
Main functional adaptations to version bit module

sanch0panza committed:
Add unit test (genversionbits_tests.cpp) which covers the new grace period functionality. The rest is still tested by versionbits_tests.cpp .

sanch0panza committed:
Changes for reading forks.csv and dumping the fork data on request

rebase: moved new help arguments into allowed_args.cpp

sanch0panza committed:
Add new 'bipgenvb_forks' output section for getblockchaininfo RPC call

sanch0panza committed:
Remove WarningBitsConditionChecker, replace with new (but primitive) signaling events logging and unknown version (51/100) alerting

sanch0panza committed:
Adapted p2p-versionbits-warning.py test to new "unknown version" detection logic

p2p-versionbits-warning needed a little adaptation after the unknown bit signaling and unknown version detection logic in main.cpp was changed.
The issuance of the "unknown version" warning has changed based on two things:
1. the threshold / window for assessing unknown bits is now 50 / 100, fixed for all networks
2. unconfigured ("unknown") bits are not assessed based on state (this is impossible since we do not know when they activate), but purely on whether they cross the 50% threshold

Some cleanup has also been performed in the file (getting rid of some hardcoded constants in favor of VB_PERIOD, VB_THRESHOLD).

sanch0panza committed:
Build system updates to include the new modules and tests

sanch0panza committed:
Documentation on the changes (still draft, as is the BIP itself of course)

sanch0panza committed:
[cleanup] Fix newly introduced compiler warnings

- missing format string in fprintf()
- moved FORKS_CSV_FILE_HEADER definition to forks_csv.cpp so that it is not reported as unused everywhere
- removed unnecessary const qualifiers on return values of validation funtions
- removed some unused variables
- surrounded the integer overflows in forkscsv_tests.cpp with GCC diagnostic pragma overrides

sanch0panza committed:
Declare explicit dependency between the test cases, to fix spurious test failure

This test is a bit naughty in that forkscsv_dumpforks_test expects a 'segwit' output for the main network.
This 'segwit' fork is configured in forkscsv_read_test, and so the test works fine as long as that case is executed first.
In my own tests this always seemed to be the case, but Travis seems to show up this dependency quite well.

The test would presumably also work without the explicit dependency if the segwit fork data were built in to the client.
That has not been done in order to keep somewhat unrelated changes in the application to a minimum.

sanch0panza committed:
[cleanup] Change from bipgenvbvoting to bip135

Renamed files whose name should reflect BIP135
Changed regtest name in RPC test runner
Replaced 'genvbvoting' with 'bip135' in file content
  (comment markers, function names, RPC interface name)
Removed some dead code from regression test

sanch0panza committed:
[cleanup] Add missing bip135_forks RPC help text

Adjust formatting of bip9_softforks "status" field to align descriptive text

sanch0panza committed:
[cleanup] Clarify sanitized pattern and set threshold to accurate 51/100

These are mostly alignments to bring it closer to the Bitcoin Core version of
this script, which eliminates some unnecessary incrementing/decrementing
by one and also has clearer use of constants.

sanch0panza committed:
[review] Use GetBoolArg to check dumpforks

sanch0panza committed:
[cleanup] Split up BIP135 regtest into separate threshold and grace period tests

This split was initially performed out of necessity while porting this PR to
Bitcoin Core, but it is better to have separate smaller tests - they are long
enough as it is. This will also make it easier to maintain synchronization
between a BU implementation and a Core implementation.

sanch0panza committed:
[review] Avoid fprintf to stdout, use C++ idioms instead

sanch0panza committed:
[review] Remove sleep from test

Not sure why it failed without these sleeps before - when I removed now it
tested ok.

sanch0panza committed:
[review] Shorten statements to single return expression

sanch0panza committed:
[review] Rename BIP9Deployment* to ForkDeployment*, and restore a comment

sanch0panza committed:
[Core review] Remove enum constants

These were probably assigned for debugging purposes.

sanch0panza committed:
[cleanup] Remove obsolete comment, rename obsolete BIP9 references to BIP135

sanch0panza committed:
[review] Factor out backAtDefined()

sanch0panza committed:
[review] Factor out warning bits check from UpdateTip into CheckAndAlertUnknownVersionbits

sanch0panza committed:
[cleanup] Format main.{h,cpp} to comply with style rules

sanch0panza committed:
Fix merge error incurred during rebase

sanch0panza committed:
Fix up some merge breakage and compilation errors due to updated logging + headers

sanch0panza committed:
Fix up the formatting

sanch0panza committed:
Remove BOM from chainparams.cpp

sickpig commented:
thanks for doing that ! 

/cc @gandrewstone

sanch0panza commented:
I need to check the basic regression tests for impacts by this PR. Could be that changes on the base coupled with these changes leads to some failing basic tests. I recall on the previous rebase prior to 1.3.0.x the basic tests still passed.

Before a merge I would also like to get bip9-softforks test to pass, which it doesn't seem to do anymore. 
One of the key points of this PR is to remain compatible with BIP9. Not sure yet what changed, but the test framework is getting a NoneType where it expects an integer.

sanch0panza commented:
my test results for basic suite right now

`TEST                           | PASSED | DURATION

tweak.py                       | True   | 3 s
may152018_forkactivation_2.py  | False  | 31 s
validateblocktemplate.py       | False  | 36 s
may152018_forkactivation_1.py  | True   | 69 s
notify.py                      | True   | 69 s
bip68-112-113-p2p.py           | True   | 161 s
listtransactions.py            | True   | 161 s
receivedby.py                  | False  | 63 s
excessive.py                   | False  | 330 s
mempool_resurrect_test.py      | True   | 33 s
txn_doublespend.py --mineblock | False  | 68 s
txn_clone.py                   | False  | 45 s
wallet.py                      | True   | 414 s
getchaintips.py                | True   | 95 s
mempool_spendcoinbase.py       | True   | 33 s
rest.py                        | True   | 127 s
rawtransactions.py             | True   | 153 s
httpbasics.py                  | True   | 14 s
mempool_reorg.py               | False  | 37 s
multi_rpc.py                   | True   | 15 s
proxy_test.py                  | True   | 14 s
parallel.py                    | True   | 628 s
signrawtransactions.py         | True   | 36 s
zapwallettxes.py               | True   | 130 s
nodehandling.py                | True   | 24 s
merkle_blocks.py               | True   | 147 s
decodescript.py                | True   | 36 s
blockchain.py                  | False  | 38 s
disablewallet.py               | True   | 3 s
fundrawtransaction.py          | False  | 238 s
keypool.py                     | True   | 21 s
invalidtxrequest.py            | True   | 43 s
reindex.py                     | True   | 191 s
abandonconflict.py             | False  | 37 s
p2p-versionbits-warning.py     | False  | 67 s
importprunedfunds.py           | True   | 86 s
sendheaders.py                 | True   | 210 s
walletbackup.py                | True   | 420 s
thinblocks.py                  | True   | 118 s

ALL                            | False  | 4444 s (accumulated)`

sickpig commented:
@sanch0panza was referring to this error in the unit tests

https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/380131087#L1506

but indeed we need to fix also the ones we find in the functional python tests. 

speaking of which `may152018_forkactivation_2.py` is unrelated to your PR, it has to do w/ the fact that we passed the protocol upgrade date so we need to change it in any case. 

`mempool_reorg.py`, `excessive.py` failing also on current `dev` w/o your changes applied on top and I guess the reason are similar to the one that cause `may152018_forkactivation_2` failure. 

So the "net" failures of this PR should be:

```
validateblocktemplate.py       | False  | 36 s
receivedby.py                  | False  | 63 s
txn_doublespend.py --mineblock | False  | 68 s
txn_clone.py                   | False  | 45 s
blockchain.py                  | False  | 38 s
fundrawtransaction.py          | False  | 238 s
abandonconflict.py             | False  | 37 s
p2p-versionbits-warning.py     | False  | 67 s
```

sanch0panza commented:
@sickpig : Thanks for the input. I re-tested the basic suite on `dev` at a655d02dfffbb933496b0f8dd7531ed02f34cc5a .

Two tests failed in my PR above that didn't fail on `dev` in my run: `walletbackup` and `p2p-versionbits-warning`. I ran those two again on the PR branch and it passed. So they are unstable.

The rest that failed in my PR also failed on `dev` without this PR, and some passed that were failed on `dev` (`getchaintips`, `mempool_spendcoinbase.py`).

Will take time to sort out where there are problems that may be linked to the PR.

sanch0panza commented:
I think the best approach for me is to try to help fix tests on `dev` as far as possible, then rebase again later, to address any defects that might be due to the PR.

ptschip commented on src/consensus/params.h:
```diff
@@ -17,9 +17,44 @@ namespace Consensus
     // NOTE: Also add new deployments to VersionBitsDeploymentInfo in versionbits.cpp
+    DEPLOYMENT_CSV = 0, // bit 0 - deployment of BIP68, BIP112, and BIP113.
+    DEPLOYMENT_SEGWIT, // bit 1 - Segregated Witness (BIP141, BIP143, BIP144)
+    // begin unassigned bits. Rename bits when allocated.
```
These bits should probably be back to unassigned?  segwit of course was never done here and bip68,112,113 has long ago activated.

sanch0panza replied:
@ptschip : I agree with you. Will submit a patch to unassign.

ptschip commented on src/test/forkscsv_tests.cpp:
```diff
@@ -0,0 +1,161 @@
+    BOOST_CHECK(ValidateForkName("a fork"));
+    BOOST_CHECK(ValidateForkName("segwit"));
+    BOOST_CHECK(!ValidateForkName(""));
+
```
maybe add a test using some special characters

sanch0panza replied:
Good idea

ptschip commented on src/versionbits.cpp:
```diff
@@ -1,21 +1,136 @@
-        /*.gbt_force =*/true,
-    }};
+        /*.name =*/ (char *) "segwit",    // bit 1, known, active but not implemented in this client
+        /*.gbt_force =*/ true,
```
is this reference to segwit still needed?

ptschip commented on doc/bip-genvbvoting-condensed-requirements.txt:
```diff
@@ -0,0 +1,123 @@
+==BIP-GENVBVOTING concise requirements specification==
+
+This is a condensed form of the requirements from BIP-GENVBVOTING.
+It is intended for checking completeness and test status.
```
I think this file should rather have an *.md extention rather than a *.txt  ?

ptschip commented on qa/rpc-tests/bip135-grace.py:
```diff
@@ -280,7 +276,6 @@ def test_BIP135GraceConditions(self):
         test_blocks = self.generate_blocks(10, 0x20000000)
         yield TestInstance(test_blocks, sync_every_block=False)
-        time.sleep(3)
         bcinfo = self.nodes[0].getblockchaininfo()
```
if you're going to remove these sleeps then shouldn't you be setting sync_every_block=true...otherwise won't you get  spurious failures on slower systems that don't sync as fast since you're assuming in the next steps that the blocks are syncd?

ptschip reviewed (changes requested):
@sanch0panza I've take a first pass over the code and will do some followup testing today and tomorrow.

ptschip commented on src/test/forkscsv_tests.cpp:
```diff
@@ -40,8 +42,11 @@ BOOST_AUTO_TEST_CASE(forkscsv_read_test)
  */
-BOOST_AUTO_TEST_CASE(forkscsv_dumpforks_test)
+BOOST_AUTO_TEST_CASE(forkscsv_dumpforks_test, * utf::depends_on("forkscsv_tests/forkscsv_read_test"))
 {
```
I'm getting a compile error here "Boost auto testcase passed 2 args but only 1 expected"...I"m wondering what version of boost you're running when you compile?  I"m on 1.58

ptschip replied:
I commented out the utf::depends_on and the tests run just fine...not sure why you need that or if it's really need it maybe you can just combine the two tests into one?

sanch0panza committed:
Squashed 'src/fast-cpp-csv-parser/' content from commit 3b439a6

git-subtree-dir: src/fast-cpp-csv-parser
git-subtree-split: 3b439a664090681931c6ace78dcedac6d3a3907e

sanch0panza committed:
Subtree import of Ben Strasser's fast CSV parser into src/fast-cpp-csv-parser

Merge commit '6acde1fbbb5a269774a5ae36f6dcd16ca66b3665' as 'src/fast-cpp-csv-parser'

Command used:
git subtree add --squash --prefix src/fast-cpp-csv-parser https://github.com/ben-strasser/fast-cpp-csv-parser master

gandrewstone commented:
I'd like to get the spec pulled into the doc directory in the BU repo as part of this PR.  Also here are 2 quotes that may be contradictory so I'd like to see clarification added to the spec:

"The threshold specifies a number of blocks, in the range of 1..windowsize, which must signal for a fork in order to lock it in. ***The support is measured when the chain height is evenly divisible by the windowsize***. If the windowsize is set to 2016 (as in BIP9) this coincides with the 2016-block re-targeting intervals."

"The signaling bits SHALL be tallied whenever the head of the active chain changes (including after reorganizations)."

sanch0panza commented:
@gandrewstone : I'll add a patch to pull in the spec, and to clarify the passages you mention.

They are not contradictory in effect - the first refers to evaluating for a fork lock-in at only those intervals determined by the windowsize. The second refers to incrementing the counts when the chain head changes. These are separate concerns, but the spec should be clearer about it.

gandrewstone commented on qa/rpc-tests/bip135-grace.py:
```diff
@@ -0,0 +1,334 @@
+                                         'active'  # locked_in
+                                         ])
+
+        # check the ones supposed to activate at next+2 period
```
The code below is the same as above... cut-paste error?

gandrewstone commented on qa/rpc-tests/bip135-grace.py:
```diff
@@ -0,0 +1,334 @@
+#!/usr/bin/env python3
```
do not use "-" in python file names.  This means they can't be used interactively:
```
$ python
>>> import bip135-grace
  File "<stdin>", line 1
    import bip135-grace
                 ^
SyntaxError: invalid syntax
```
interactive use is nice for development and debugging.

sanch0panza replied:
ok, files will be renamed

gandrewstone commented on qa/rpc-tests/bip135-grace.py:
```diff
@@ -0,0 +1,334 @@
+            print(info['bip135_forks'][f])
+
+    def test_BIP135GraceConditions(self):
+        print("begin test_BIP135GraceConditions test")
```
please convert prints to logging.  This cuts down the size of the already-huge travis output.  At the top:
import logging
logging.basicConfig(format='%(asctime)s.%(levelname)s: %(message)s', level=logging.INFO, stream=sys.stdout)

then:
logging.info("begin test_BIP135GraceConditions test")

ftrader replied:
Thanks will do

sanch0panza replied:
will do, thanks

gandrewstone commented on qa/rpc-tests/bip135-threshold.py:
```diff
@@ -0,0 +1,317 @@
+#!/usr/bin/env python3
```
rename without "-" and use logging in this file too, thx!

sanch0panza replied:
ok, np

gandrewstone commented on qa/rpc-tests/bip135-threshold.py:
```diff
@@ -0,0 +1,317 @@
+            "regtest,6,bip135test6,%d,999999999999,100,99,0,0,true\n" % (self.fork_starttime) +
+
+            # bit 7: max threshold
+            "regtest,7,bip135test7,%d,999999999999,100,100,0,0,true\n" % (self.fork_starttime)
```
I would like to see a test that verifies:
 
1. the timeout.  The test should produce threshold-1 blocks in windowsize before timeout, and then produces a block with the bit set after timeout.  This should NOT activate (look for self.nodes[0].setmocktime(time) in other tests to see how to manipulate time in your test).

2. Previously I mentioned confusion as to when activation bits were counted.  If they are counted only when blockheight%windowsize==0, I would like to see a test that produces more than threshold blocks with the bit set, but does NOT activate because these blocks cross a windowsize boundary.

3. What edge conditions did I miss?

We need to test these specification edge conditions in the python code because the python code can easily be used for interoperability testing (between full nodes), whereas the C code is quite locked into a specific implementation.

gandrewstone commented on src/bitcoind.cpp:
```diff
@@ -18,6 +19,7 @@
 #include "rpc/server.h"
 #include "scheduler.h"
 #include "unlimited.h"
+#include "unlimited.h"
```
double include

sanch0panza replied:
thanks, will remove

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -112,16 +113,20 @@ class CMainParams : public CChainParams
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].windowsize = 2016;
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].threshold = 1916; // 95% of 2016
+
+        // testing bit
```
why did the order get swapped?  It'll be a cleaner diff without swapping

ftrader replied:
ok

sanch0panza replied:
not sure, will align to make diff cleaner

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -404,14 +409,18 @@ class CRegTestParams : public CChainParams
-        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nTimeout = 999999999999ULL;
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].nTimeout = 999999999999LL;
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].windowsize = 144;
+        consensus.vDeployments[Consensus::DEPLOYMENT_CSV].threshold = 108; // 75% of 144
```
order swap

sanch0panza replied:
ok

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -487,3 +496,68 @@ void SelectParams(const std::string &network)
+ */
+bool isConfiguredDeployment(const Consensus::Params &consensusParams, const int bit)
+{
+    const Consensus::ForkDeployment *vdeployments = consensusParams.vDeployments;
```
range check bit >=0 and < MAX_VERSION_BITS_DEPLOYMENTS, DbgAssert(predicate, return false)  to catch in debug mode but continue in release.

gandrewstone commented on src/consensus/params.h:
```diff
@@ -17,23 +17,68 @@ namespace Consensus
+    DEPLOYMENT_UNASSIGNED_BIT_27,
+    DEPLOYMENT_TESTDUMMY, // bit 28 - used for deployment testing purposes
+    // bip135 end
     MAX_VERSION_BITS_DEPLOYMENTS
```
semantically, the "max" should be inclusive, so 28.  "What's the maximum score on this test?", you don't answer "101" for the typical 0-100 test score range.  So what is defined as MAX_VERSION_BITS_DEPLOYMENTS, should really be NUM_VERSION_BITS_DEPLOYMENTS -- the number of version bits is 29

Please change the name if you agree, but I'm not going to block the PR for this triviality...

gandrewstone replied:
Also, what is the relationship between this and VERSIONBITS_NUM_BITS?  If they are the same, please consolidate. 1 idea should have exactly 1 constant to help code maintenance.  Otherwise you could have a case where only 1 of the 2 constants gets changed.

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -487,3 +496,68 @@ void SelectParams(const std::string &network)
+ * Unconfigured deployments can be ignored to save processing time, e.g.
+ * in ComputeBlockVersion() when computing the default block version to emit.
+ */
+bool isConfiguredDeployment(const Consensus::Params &consensusParams, const int bit)
```
Capitalize first letter of function names:
"IsConfiguredDeployment"

sanch0panza replied:
ok, thanks

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -487,3 +496,68 @@ void SelectParams(const std::string &network)
+const std::string NetworkDeploymentInfoCSV(const std::string &network)
+{
+    const Consensus::Params &consensusParams = Params(network).GetConsensus();
+    ;
```
extra ;

gandrewstone commented on src/chainparams.h:
```diff
@@ -70,6 +70,8 @@ class CChainParams
     };
 
     const Consensus::Params &GetConsensus() const { return consensus; }
+    /** Modifiable consensus parameters added by bip135 */
```
add "this can only be used during initialization, since modification of the consensus params is not thread safe" to the comment.

gandrewstone commented on src/chainparams.h:
```diff
@@ -139,4 +141,24 @@ CBlock CreateGenesisBlock(CScript prefix,
+/**
+ * Return the currently selected parameters. Can be changed by reading in
+ * some additional config files (e.g. CSV deployment data)
+ */
```
add same as above to the docs of this function too.

gandrewstone commented on src/consensus/params.h:
```diff
@@ -17,23 +17,68 @@ namespace Consensus
+    /** Minimum number of blocks to remain in locked-in state */
+    int minlockedblocks;
+    /** Minimum duration (in seconds based on MTP) to remain in locked-in state */
+    int64_t minlockedtime;
```
does a negative time have meaning?

gandrewstone commented on src/init.cpp:
```diff
@@ -900,6 +905,49 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
+    else
+    {
+        // be noisy, but don't fail if file is absent - use built-in defaults
+        LOGA("No deployment configuration found at '%s' - using defaults\n", ForksCsvFile);
```
If the user configured a fork file but we can't find it, I think that we should exit with an error.  

If the default fork file is not there, its ok to log and continue.

gandrewstone commented on src/rpc/blockchain.cpp:
```diff
@@ -742,6 +742,40 @@ static UniValue BIP9SoftForkDesc(const Consensus::Params &consensusParams, Conse
+// bip135 begin
+static UniValue BIP135ForkDesc(const Consensus::Params &consensusParams, Consensus::DeploymentPos id)
+{
+    UniValue rv(UniValue::VOBJ);
```
range check "id"

gandrewstone commented on src/rpc/blockchain.cpp:
```diff
@@ -771,7 +805,7 @@ UniValue getblockchaininfo(const UniValue &params, bool fHelp)
             "  \"bip9_softforks\": {          (object) status of BIP9 softforks in progress\n"
             "     \"xxxx\" : {                (string) name of the softfork\n"
-            "        \"status\": \"xxxx\",    (string) one of \"defined\", \"started\", \"lockedin\", \"active\", "
+            "        \"status\": \"xxxx\",      (string) one of \"defined\", \"started\", \"lockedin\", \"active\", "
```
are these addtl spaces intentional?

gandrewstone commented on src/versionbits.cpp:
```diff
@@ -1,21 +1,141 @@
     },
     {
-        /*.name =*/"csv",
+        /*.name =*/(char *)"segwit", // bit 1, known, active but not implemented in this client
```
remove segwit

gandrewstone commented on src/versionbits.cpp:
```diff
@@ -24,17 +144,29 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex
     // whose height equals a multiple of nPeriod - 1.
     if (pindexPrev != NULL)
     {
+        assert(nPeriod);
```
this assertion is guaranteed true due to the previous if statement

gandrewstone commented on src/versionbits.cpp:
```diff
@@ -24,17 +144,29 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex
     }
 
-    // Walk backwards in steps of nPeriod to find a pindexPrev whose information is known
+    // Walk backwards in steps of nPeriod to find a pindexPrev which was DEFINED
```
check pindexPrev for NULL, if its null, some garbage was put in the fork definition, like a period larger than the current block height.  Return THRESHOLD_DEFINED?
EDIT: Nevermind you check null below

gandrewstone commented:
Let me know when you've addressed the comments -- I want to get the merged in before new conflicts occur.
